### ISSUE TITLE
Can I use fp8 only when the code runs to the fp8 branch?

### ISSUE THREAD

**Issue by forevergj** [Azure/MS-AMP#176]
I am using deepspeed. ops. adam FusedAdam cannot go to the fp8 branch
![image](https://github.com/Azure/MS-AMP/assets/43092496/b73dc694-2217-42d8-b5c4-55fe77a063e2)
deepspeed zero stage2

**Comment by wkcn** [comment]
Hi @forevergj , thanks for your attention to our work!

The reason is that master weight, weight, weight gradient and optimizer states are all tensors with scaling factors. FusedAdam did not support the computation on scaling tensors.

**Comment by forevergj** [comment]
Thank you for your reply！Where do the benefits of using msamp for training acceleration come from?

**Comment by wkcn** [comment]
@forevergj 
MS-AMP applies low-bit data formats on master weight, weight, weight gradient and optimizer states. It saves the GPU memory to allow larger batch size for faster training speed.

**Comment by forevergj** [comment]
May I ask how to reproduce the results in the paper？ I used two H800GPUs.
msamp: samples per second: 8.130 | TFLOPs: 356.36
bf16：samples per second: 8.8| TFLOPs: 356.36
 I run msamp and bf16 on https://github.com/Azure/MS-AMP-Examples GPT 6.7B, msamp has decreased performance compared to bf16.with this：
`#!/bin/bash

# Runs the "6.7b" parameter model
# GPT-13B: 32 layers, 4096 hidden size, 32 attention heads

set -e

USAGE="usage: bash pretrain_6b7_megatron.sh [bf16|te|msamp]"
export CUDA_VISIBLE_DEVICES=6,7
if [ "$#" -ne 1 ]; then
  echo $USAGE
  exit 1
fi

FP_TYPE=$1

export CUDA_DEVICE_MAX_CONNECTIONS=1
GPUS_PER_NODE=2
# Change for multinode config
MASTER_ADDR=localhost
MASTER_PORT=6000
NNODES=1
NODE_RANK=0
WORLD_SIZE=$(($GPUS_PER_NODE*$NNODES))

VOCAB_FILE=$PWD/data/gpt2-vocab.json
MERGE_FILE=$PWD/data/gpt2-merges.txt
DATA_PATH=$PWD/data/wikipedia_text_document

DISTRIBUTED_ARGS="
    --nproc_per_node $GPUS_PER_NODE \
    --nnodes $NNODES \
    --node_rank $NODE_RANK \
    --master_addr $MASTER_ADDR \
    --master_port $MASTER_PORT
"

GPT_ARGS="
    --tensor-model-parallel-size 1 \
    --pipeline-model-parallel-size 2 \
    --distributed-backend nccl \
    --no-query-key-layer-scaling \
    --seed 43 \
    --num-layers 32 \
    --hidden-size 4096 \
    --num-attention-heads 32 \
    --seq-length 2048 \
    --max-position-embeddings 2048 \
    --train-samples 48828125 \
    --lr-decay-samples 43945312  \
    --lr-warmup-samples 2048000  \
    --lr 3.0e-4 \
    --min-lr 3.0e-5 \
    --lr-decay-style cosine \
    --micro-batch-size 1 \
    --global-batch-size 2048 \
    --clip-grad 1.0 \
    --weight-decay 0.1 \
    --attention-dropout 0.0 \
    --hidden-dropout 0.0 \
    --optimizer adam \
    --adam-beta1 0.9 \
    --adam-beta2 0.95 \
    --init-method-std 0.0099 \
    --num-workers 1 \
    --bf16 \
    --sequence-parallel \
    --use-flash-attn \
    --no-gradient-accumulation-fusion \
    --use-distributed-optimizer
"


DATA_ARGS="
    --data-path $DATA_PATH \
    --vocab-file $VOCAB_FILE \
    --merge-file $MERGE_FILE \
    --data-impl mmap \
    --split 949,50,1
"

OUTPUT_ARGS="
    --log-interval 1 \
    --save-interval 1000 \
    --eval-interval 500 \
    --eval-iters 7
"

if [ "$FP_TYPE" = "bf16" ]; then
    CHECKPOINT_PATH=$PWD/checkpoints/gpt_6b7_bf16
    torchrun $DISTRIBUTED_ARGS ../third_party/Megatron-LM/pretrain_gpt.py \
        $GPT_ARGS \
        $DATA_ARGS \
        $OUTPUT_ARGS \
        --save $CHECKPOINT_PATH \
        --load $CHECKPOINT_PATH
elif [ "$FP_TYPE" = "te" ]; then
    CHECKPOINT_PATH=$PWD/checkpoints/gpt_6b7_te
    torchrun $DISTRIBUTED_ARGS ../third_party/Megatron-LM/pretrain_gpt.py \
        $GPT_ARGS \
        $DATA_ARGS \
        $OUTPUT_ARGS \
        --save $CHECKPOINT_PATH \
        --load $CHECKPOINT_PATH \
        --fp8-hybrid \
        --transformer-impl transformer_engine
elif [ "$FP_TYPE" = "msamp" ]; then
    CHECKPOINT_PATH=$PWD/checkpoints/gpt_6b7_msamp
    torchrun $DISTRIBUTED_ARGS ../third_party/Megatron-LM/pretrain_gpt.py \
        $GPT_ARGS \
        $DATA_ARGS \
        $OUTPUT_ARGS \
        --save $CHECKPOINT_PATH \
        --load $CHECKPOINT_PATH \
        --fp8-hybrid \
        --transformer-impl transformer_engine \
        --msamp
else
    echo $USAGE
    exit 1
fi
`

**Comment by wkcn** [comment]
> > MS-AMP applies low-bit data formats on master weight, weight, weight gradient and optimizer states. It saves the GPU memory to allow larger batch size for faster training speed.
> 
> 1.Does it mean that the fp8 operation of msamp only occurs after backpropagation when the optimizer updates the weights. However, forward propagation does not involve fp8 operations? 2.Is the main optimization focus on gradient communication time?

1. The forward propagation and backward propagation both involve FP8.
2. No. The acceleration benefits from FP8 matrix multiplication in the linear layers. MS-AMP reduces the GPU memory by low-precision data formats. It allows larger batch size for the further acceleration.

**Comment by forevergj** [comment]
![image](https://github.com/Azure/MS-AMP/assets/43092496/7275f19d-1029-43fd-808a-26ea8343227e)
Is it similar to a transformer engine? What is the difference between msamp and transformer engine？

**Comment by wkcn** [comment]
> ![image](https://private-user-images.githubusercontent.com/43092496/330638920-7275f19d-1029-43fd-808a-26ea8343227e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU3NDM1NDcsIm5iZiI6MTcxNTc0MzI0NywicGF0aCI6Ii80MzA5MjQ5Ni8zMzA2Mzg5MjAtNzI3NWYxOWQtMTAyOS00M2ZkLTgwOGEtMjZlYTgzNDMyMjdlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE1VDAzMjA0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExOGFkNDljMjU0YTM3N2MxNzk4MGI5ZDlmYTI2YjI2MTE4ODRhNTY0NDAzZjg1MDk3MDAxY2I5NGQyM2ZlYTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XiIZTCukN8F-H7ffg8-T5wcWEKSHOygQ5V0TfVf2k50) Is it similar to a transformer engine? What is the difference between msamp and transformer engine？ ![image](https://private-user-images.githubusercontent.com/43092496/330642306-9564282e-ac36-47a3-b35c-9e377b9e149f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU3NDM1NDcsIm5iZiI6MTcxNTc0MzI0NywicGF0aCI6Ii80MzA5MjQ5Ni8zMzA2NDIzMDYtOTU2NDI4MmUtYWMzNi00N2EzLWIzNWMtOWUzNzdiOWUxNDlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE1VDAzMjA0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkNjA0MjNiYmQ2NmQ1ZmY3NjZiODgzY2RhYmQ2YzE5OGMxMjE2NWM2N2JiMTViYTUyN2UyOWYxZWVhNmI1NzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DvqAMH_o6VhS-DuE3GIsw_HOn2R05of0QotfPIJKogM) Does the image in the MSAMP paper mean that the conversion of FP8 is executed in two places. They are before the all-gather gradient and before executing Linear. This means that the output of Linear is high-precision？

1. MS-AMP can be combined with TransformerEngine. The difference is that MS-AMP applies FP8 data format on master weight, weight, weight gradient, optimizer states, and the communication of gradient reduction.

2. Regarding to your second question, the output of LayerNorm is a high-precision tensor (FP16 or BF16), which is converted to an FP8 tensor with a scaling factor, as the input of the all-gather operation. The output of the all-gather operation is an FP8 tensor with a scaling factor, that is the input of the FP8 GEMM (`YA`). The output of the FP8 GEMM (`YA`) is a high-precision tensor (FP16 or BF16). Z is also a high-precision tensor, being converted to an FP8 tensor with a scaling factor.

**Comment by forevergj** [comment]
Thank you for your reply. It's very clear
